### PR TITLE
Experimenting with different authentication methods

### PR DESCRIPTION
**Pull request by MokhtarAkle**
# Pull request 05-02-2024 - 04-03-2024

<details>
<summary>
Summary
</summary>  
  
At the moment I'm experimenting with different auth methods to implement a functional login system. My choices have been made based on available tutorials/ ease of use as I am a beginner when it comes to auth in web dev. As the most recommended auth library, Lucia is written in typescript, I've added TS to my project when needed. I've also added a local db and some minor changes to the project structure that shouldn't affect the rest of the project.
</details>

<details>
<summary>
Related issues
</summary>  
 #86 #88 #89 #90 
</details>

image

 **Link: https://grote-themas-sprint-19-rosy.vercel.app/**

Fixes: Nothing

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] User test
- [ ] Accessibility test
- [ ] Performance test
- [ ] Responsive Design test

# Checklist:

- [x] My code follows the style guidelines of this project
- [ ] The code is accessible
- [ ] The code is performant
- [ ] The code is responsive
- [x] I have written meaningful commit messages
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (README and/or Wiki)

# Pull request 04-03-2024 - 06-03-2024

<details>
<summary>
Summary
</summary>  
  
Since my last pull request I've decided on an authentication method to move forward with, Lucia. I have also switched ORM to Drizzle from Prisma after finding more resources related to my specific tech stack. After comitting to this stack and finding  different tutorials and documentation I've managed to add a working registration function that stores a user_session and user object in the local db. The inputs used are validated through a combination of superforms (form library for svelte) and zod. After this data is stored I've managed to retrieve it and check it against incoming data from a different input and when entering a route. This means a basic login functionality has been added and there are routes that can't be accessed without being logged in. I also added a simple logout function to the different pages that disposes of a session when clicked and removes the cookie before redirecting you.

</details>

<details>
<summary>
Related issues
</summary>  
 #86 #88 #89 #90 #93
</details>

image

 **Link: https://grote-themas-sprint-19-rosy.vercel.app/**

Fixes: Site not starting, zod not validating input. logout action not working on certain pages, buttons not triggering functions.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] User test
- [ ] Accessibility test
- [ ] Performance test
- [ ] Responsive Design test

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] The code is accessible
- [ ] The code is performant
- [ ] The code is responsive
- [x] I have written meaningful commit messages
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README and/or Wiki)

# Pull request 06-03-2024 - 14-03-2024

<details>
<summary>
Summary
</summary>  

Since adding the basic functionalities of the whole authentication systems, I've been working on moving from a locally hosted db to a server hosted db so the authentication can be used in a live environment. After looking through my available options and asking a lecturer, I decided to move forward with a vercel postgres db. Thankfully Drizzle has a good migration system to transfer schemas to different db's but I unfortunately met with some nasty bugs. After squishing those I decided to finalize the authentication system by adding some front-end styling for the authentication components.

</details>

<details>
<summary>
Related issues
</summary>  
#91 #92 #94 #95 #97 #98
</details>

image

 **Link: https://grote-themas-sprint-19-rosy.vercel.app/**

Fixes: Timestamp type not being cast to integer, site not deploying due to local db and casing inconsistencies, .env variables not being recognized.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] User test
- [x] Accessibility test
- [ ] Performance test
- [x] Responsive Design test

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] The code is accessible
- [ ] The code is performant
- [x] The code is responsive
- [x] I have written meaningful commit messages
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README and/or Wiki)

### CODE REVIEW

**Comment by joostf** [comment]
@MokhtarAkle please discuss login strategy with @ju5tu5!

**Comment by MokhtarAkle** [comment]
> @MokhtarAkle please discuss login strategy with @ju5tu5!

I had a small discussion with him and we agreed that the method should be decided by the assignment giver. Until I get in contact with him and hear his wishes I'll be continuing with this method for personal experience. In the next PR the method will be working as well so it seems to be a valid option!

**Comment by ju5tu5** [comment]
We talked about this indeed. I mentioned that i prefer magic link signin, especially because this doesn't have a difficult user management system. Of course, as you mention, the preference of your client is mandatory...

**Comment by MokhtarAkle** [comment]
> We talked about this indeed. I mentioned that i prefer magic link signin, especially because this doesn't have a difficult user management system. Of course, as you mention, the preference of your client is mandatory...

From my (limited) experience the current method does not have a particulary difficult user management system seeing as I only store some basic user data and a session ID to validate the login. Of course I'm still planning to implement magic links as discussed although it'd probably be in a seperate repo for now so I don't start mixing things. Hopefully I'll be able to present both to the client for an easier choice.

File: drizzleMigrations/meta/0000_snapshot.json
Comment (ju5tu5) [inline drizzleMigrations/meta/0000_snapshot.json:1]: Horen deze snapshots in de PR?

File: drizzleMigrations/meta/_journal.json
Comment (ju5tu5) [inline drizzleMigrations/meta/_journal.json:1]: Same here.. dit lijkt dummy data.. moet dat in een live omgeving?

File: package.json
Comment (ju5tu5) [inline package.json:1]: Ik zie drizzle (voor SQL right?) en MongoDB.. gebruik je 2 databases? Zo niet kan mongo weg denk ik...

File: src/lib/atoms/NavFilterList.svelte
Comment (ju5tu5) [inline src/lib/atoms/NavFilterList.svelte:1]: Beter...

File: src/lib/atoms/SuperValidButton.svelte
Comment (ju5tu5) [inline src/lib/atoms/SuperValidButton.svelte:1]: Dit lijkt me meer een molecule.. omdat het een atom gebruikt..

File: src/lib/atoms/inputfields/SelectField.svelte
Comment (ju5tu5) [inline src/lib/atoms/inputfields/SelectField.svelte:1]: Same.. meer een molecule.. er worden immers meerdere elementen genest...

File: src/lib/atoms/inputfields/SuperValidInput.svelte
Comment (ju5tu5) [inline src/lib/atoms/inputfields/SuperValidInput.svelte:1]: Duidelijk molecule..

File: src/lib/atoms/specialButton/button.svelte
Comment (ju5tu5) [inline src/lib/atoms/specialButton/button.svelte:1]: ...molecule...

File: src/lib/database/auth.server.ts
Comment (ju5tu5) [inline src/lib/database/auth.server.ts:1]: Hoe veilig is lucia? And again.. je verplicht je opdrachtgever een database te onderhouden met namen en emailadressen. Het is echt veel handiger om magic-link authenticatie te implementeren...

File: src/lib/organisms/UploadForm.svelte
Comment (ju5tu5) [inline src/lib/organisms/UploadForm.svelte:1]: Wat mij betreft is dit bestand veeel te groot. Probeer het op te splitsen in logische atomic onderdelen in je lib..

File: src/lib/validations/AuthZodSchemas.ts
Comment (ju5tu5) [inline src/lib/validations/AuthZodSchemas.ts:1]: De minimum lengte's moeten omhoog, dit is op dit moment niet veilig..

File: src/routes/[id]/+page.svelte
Comment (ju5tu5) [inline src/routes/[id]/+page.svelte:1]: Zet inlined SVG's ook in je lib..

Comment (ju5tu5) [reply src/routes/[id]/+page.svelte:1]: Bestand is ook best wel lang...

File: svelte.config.js
Comment (ju5tu5) [inline svelte.config.js:1]: Mooie aliases! Goed dat je die toevoegt..

**Review by ju5tu5 — CHANGES_REQUESTED** [review]
Je code stijl is best oke. Ik twijfel of Lucia definitief de juiste authenticatie is maar dat is een lopende discussie, het werkt voor nu. Je bestanden hebben de neiging nogal groot te worden en atomic design is niet overal even netjes toegepast, zou je daar nog eens naar kunnen kijken? Ik heb hier en daar comments toegevoegd om wat ik bedoel te verduidelijken.

File: drizzleMigrations/meta/0000_snapshot.json
Comment (MokhtarAkle) [reply drizzleMigrations/meta/0000_snapshot.json:1]: Nee zeker niet, drizzleMigrations is toegevoegd aan .gitignore en verwijderd uit de repo

File: drizzleMigrations/meta/_journal.json
Comment (MokhtarAkle) [reply drizzleMigrations/meta/_journal.json:1]: Nee zeker niet, drizzleMigrations is toegevoegd aan .gitignore en verwijderd uit de repo

File: package.json
Comment (MokhtarAkle) [reply package.json:1]: Alle ongebruikte dependencies zijn verwijderd en package.json en package-lock.json

File: src/lib/atoms/SuperValidButton.svelte
Comment (MokhtarAkle) [reply src/lib/atoms/SuperValidButton.svelte:1]: Veranderd naar molecule

File: src/lib/atoms/inputfields/SelectField.svelte
Comment (MokhtarAkle) [reply src/lib/atoms/inputfields/SelectField.svelte:1]: veranderd naar molecule

File: src/lib/atoms/inputfields/SuperValidInput.svelte
Comment (MokhtarAkle) [reply src/lib/atoms/inputfields/SuperValidInput.svelte:1]: veranderd naar molecule

File: src/lib/atoms/specialButton/button.svelte
Comment (MokhtarAkle) [reply src/lib/atoms/specialButton/button.svelte:1]: veranderd naar molecule en haal icon op van atom

File: src/lib/database/auth.server.ts
Comment (MokhtarAkle) [reply src/lib/database/auth.server.ts:1]: Over veiligheid kan ik niet heel veel zeggen maar gebaseerd op internet resultaten is het op z'n minst veilig om te gebruiken als dev. De developer ervan is ook erg actief en commit bijna elke dag naar de Lucia repo. Na gesprek met de opdrachtgever snappen we beide niet wat voor onderhoud er nodig is voor de database, uitleg zou gewardeerd zijn. Magic-link staat nog steeds in de planning als een could have dus als er tijd voor is zal ik het proberen te implementeren.

File: src/lib/organisms/UploadForm.svelte
Comment (MokhtarAkle) [reply src/lib/organisms/UploadForm.svelte:1]: Validator uit de uploadform gehaald en in eigen component gestopt, verder splitsing lijkt mij niet logisch maar hoor graag feedback hierop.

File: src/lib/validations/AuthZodSchemas.ts
Comment (MokhtarAkle) [reply src/lib/validations/AuthZodSchemas.ts:1]: Na advies van Cyd de minimum naam lengte verlaagd naar 1 en de minimum wachtwoord lengte verhoogd naar 10

File: src/routes/[id]/+page.svelte
Comment (MokhtarAkle) [reply src/routes/[id]/+page.svelte:1]: Inlined SVG verwijderd en img met src toegevoegd, zal nog even kijken naar het bestand voor mogelijke componenten

File: svelte.config.js
Comment (MokhtarAkle) [reply svelte.config.js:1]: Bedankt! Zal er voor zorgen dat ze overal gebruikt worden waar dit logisch is!

File: src/routes/login/+page.svelte [hunk 1]
```diff
@@ -1,59 +1,102 @@
...
-	/>
-
-	<SuperValidButton></SuperValidButton>
-</form>
\ No newline at end of file
+<main>
+
+
+  <form method="post" use:enhance action={route("logInUser /login")}>
+	<section>
```
Comment (RayanSp) [inline src/routes/login/+page.svelte:42]: Je zou ipv een section en een h1, gebruik kunnen maken van een fieldset met daarin een legend.

**Review by RayanSp — COMMENTED** [review]
Code ziet er voor zo ver ik kan zien prima uit.

**Pull request merged**